<a href="https://colab.research.google.com/github/AkshayGawande07/Python/blob/master/01PracticeQuestionsOnFunctions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1. Write a function that inputs a number and prints the multiplication table of that number

In [1]:
def table(num):
    for i in range(1,11):
        print(i*num)
        
table(17)

17
34
51
68
85
102
119
136
153
170


2. Write a program to print twin primes less than 1000. If two consecutive odd numbers are both prime then they are known as twin primes

In [3]:
def isTwinPrime(primeNumber):
    print('List of Twin Prime Number:')
    for i in range(len(primeNumber)-1):
        if (abs(primeNumber[i]-primeNumber[i+1])==2):
            print(primeNumber[i],primeNumber[i+1])
            
def isPrime():
    primeNumber = []
    for num in range(1, 1001):
        if num > 1:
            for i in range(2, num):
                if (num % i) == 0:
                    break
            else:
                primeNumber.append(num)
                
    isTwinPrime(primeNumber)


isPrime()

List of Twin Prime Number:
3 5
5 7
11 13
17 19
29 31
41 43
59 61
71 73
101 103
107 109
137 139
149 151
179 181
191 193
197 199
227 229
239 241
269 271
281 283
311 313
347 349
419 421
431 433
461 463
521 523
569 571
599 601
617 619
641 643
659 661
809 811
821 823
827 829
857 859
881 883


3. Write a program to find out the prime factors of a number. Example: prime factors of 56 - 2, 2, 2, 7

In [5]:
def findFactors(num):
    factors = []
    
    while (num%2==0):
        factors.append(2)
        num/=2
        
    for i in range (3,int(num)):
        if(num%i==0):
            factors.append(i)
            num/=i
            
    if num>2:
        factors.append(int(num))
        
    print(factors)
    
findFactors(7429)

[17, 19, 23]


4. Write a program to implement these formulae of permutations and combinations. Number of permutations of n objects taken r at a time: p(n, r) = n! / (n-r)!. Number of combinations of n objects taken r at a time is: c(n, r) = n! / (r!*(n-r)!) = p(n,r) / r!

In [6]:
def functionName(n,r):
    nFactorial = 1
    rFactorial = 1
    nrFactorial = 1
    
    nFactorial = calculateFactorial(n)
    rFactorial = calculateFactorial(r)
    nrFactorial = calculateFactorial(n-r)
    
    pnr=nFactorial/nrFactorial
    cnr=nFactorial/(rFactorial*nrFactorial)
    
    print('PNR : ',int(pnr))
    print('CNR : ',int(cnr))
    
def calculateFactorial(num):
    fact = 1
    for i in range (1,num+1):
        fact*=i
    return fact
    
functionName(6,2)
    

PNR :  30
CNR :  15


5. Write a function that converts a decimal number to binary number

In [10]:
def decimalToBinary(n):
    if n > 1:
        decimalToBinary(n//2)
    print(n % 2,end = '')

decimalToBinary(12)
print()

1100


6. Write a function cubesum() that accepts an integer and returns the sum of the cubes of individual digits of that number. Use this function to make functions PrintArmstrong() and isArmstrong() to print Armstrong numbers and to find whether is an Armstrong number.

In [11]:
def cubeSum(num):
    num=str(num)
    vSum = 0
    for i in num:
        vSum+=(int(i)**3)
        
    printArmStrong(vSum,num)

def printArmStrong(vSum,num):
    if isArmstrong(vSum,num):
        print('Is Armstrong :', num)
        print('With Cube Sum :', vSum)
    else:
        print('is NOT Armstrong :', num)
        print('With Cube Sum :', vSum)
    
def isArmstrong(vSum,num):
    if (int(vSum)==int(num)):
        return True
    else:
        return False
    
cubeSum(408)

is NOT Armstrong : 408
With Cube Sum : 576


7. Write a function prodDigits() that inputs a number and returns the product of digits of that number.

In [12]:
def prodDigits(num):
    num = str(num)
    prod = 1
    for i in num:
        prod*=int(i)
    return (int(prod))
    
prod = prodDigits(564879)
print(prod)

60480


8. If all digits of a number n are multiplied by each other repeating with the product, the one digit number obtained at last is called the multiplicative digital root of n. The number of times digits need to be multiplied to reach one digit is called the multiplicative persistance of n.<br>
 Example: <br>
     86 -> 48 -> 32 -> 6 (MDR 6, MPersistence 3)
     <br>341 -> 12->2 (MDR 2, MPersistence 2)
     <br>
Using the function prodDigits() of previous exercise write functions MDR() and
MPersistence() that input a number and return its multiplicative digital root and
multiplicative persistence respectively

In [14]:
def MDR(num):
    while (len(str(num))>1):
        prod=prodDigits(int(num))
        num = prod
    print('MDR :',num)
    
def MPersistence(num):
    c=0
    while(len(str(num))>1):
        prod=prodDigits(int(num))
        num = prod
        c+=1
    print('MPersistence :',c)

MDR(86)
MPersistence(86)

MDR : 6
MPersistence : 3


9. Write a function sumPdivisors() that finds the sum of proper divisors of a number. Proper divisors of a number are those numbers by which the number is divisible, except the number itself. For example proper divisors of 36 are 1, 2, 3, 4, 6, 9, 18

In [15]:
def sumPdivisors(num):
    divisors = []
    for i in range (1,num):
        if (num%i==0):
            divisors.append(i)
    return (sum(divisors))

sumPdiv=sumPdivisors(220)
print(sumPdiv)

284


10. A number is called perfect if the sum of proper divisors of that number is equal to the number. For example 28 is perfect number, since 1+2+4+7+14=28. Write a program to print all the perfect numbers in a given range

In [16]:
def perfectNumber(num1,num2):
    for i in range(num1,num2+1):
        sumPDiv = sumPdivisors(i)
        if (i==sumPDiv):
            print('Perfect Number : ',i)

perfectNumber(28,29)

Perfect Number :  28


11. Two different numbers are called amicable numbers if the sum of the proper divisors of each is equal to the other number. For example 220 and 284 are amicable numbers.

In [17]:
def amicableNumber(num1,num2):
    sNum1 = sumPdivisors(num1)
    sNum2 = sumPdivisors(num2)
    
    if (sNum1==num2 and sNum2==num1):
        print('{0} and {1} are Amicabel Numbers.'.format(num1,num2))
    else:
        print('{0} and {1} are NOT Amicabel Numbers.'.format(num1,num2))

amicableNumber(1184,1210)

1184 and 1210 are Amicabel Numbers.


12. Write a program which can filter odd numbers in a list by using filter function

In [18]:
def oddNumber(num):
    if (num%2==0):
        return False
    else:
        return True

numberList = [1,2,3,4,5,6,7,8,9,10,11,12,13,99,98,97,96,95,94]    
oddNumber = filter(oddNumber,numberList)

print('Odd Numbers:')
print(*(i for i in oddNumber), sep=', ', end='')

Odd Numbers:
1, 3, 5, 7, 9, 11, 13, 99, 97, 95

13. Write a program which can map() to make a list whose elements are cube of elements in a given list

In [19]:
def calculateCube(num):
    return num**3

numList = [1,2,3,4,5,6,7,8,9]
x = map(calculateCube, (numList))
print('Cube of Elements')
print(list(x))

Cube of Elements
[1, 8, 27, 64, 125, 216, 343, 512, 729]


14. Write a program which can map() and filter() to make a list whose elements are cube of even number in a given list

In [20]:
#Reference source : https://www.geeksforgeeks.org/python-lambda-anonymous-functions-filter-map-reduce/

number = [1,2,3,4,5,6,7,8,9,10,11,12,13,99,98,97,96,95,94]    
 
evenCube = map(lambda x: x**3, filter(lambda   x: x%2==0, number))
 
print(list(evenCube))

[8, 64, 216, 512, 1000, 1728, 941192, 884736, 830584]
